core/src/main/clojure/xtdb/buffer_pool.clj

```diff
@@ -116,18 +146,74 @@
+
+(defrecord LocalBufferPool [allocator, ^MemoryCache memory-cache, ^Path disk-store, ^Cache arrow-footer-cache]
   IBufferPool
-  (getBuffer [_ k]
+  (getByteBuffer [_ k]
     (when k
-      (.get memory-cache k
-            (fn [^Path k]
-              (let [buffer-cache-path (.resolve disk-store k)]
-                (when-not (util/path-exists buffer-cache-path)
-                  (throw (os/obj-missing-exception k)))
-
-                (CompletableFuture/completedFuture
-                 (Pair. buffer-cache-path nil)))))))
+      (util/with-open [arrow-buf (.get memory-cache (PathSlice. k)
+                                       (fn [^PathSlice path-slice]
+                                         (let [buffer-cache-path (.resolve disk-store (.getPath path-slice))]
+                                           (when-not (util/path-exists buffer-cache-path)
+                                             (throw (os/obj-missing-exception k)))
+
+                                           (CompletableFuture/completedFuture
+                                            (Pair. (->resolved-path-slice path-slice buffer-cache-path) nil)))))]
+        (copy-bb (.nioBuffer arrow-buf 0 (.capacity arrow-buf))))))
+
+  (getFooter [_ k]
+    (let [path (.resolve disk-store k)]
+      (when-not (util/path-exists path)
+        (throw (os/obj-missing-exception path)))
+
+      (.get arrow-footer-cache k (fn [_] (Relation/readFooter (path->seekable-byte-channel path))))))
+
+  (getRecordBatch [_ k block-idx]
+    (let [path (.resolve disk-store k)]
+      (when-not (util/path-exists path)
+        (throw (os/obj-missing-exception path)))
+
+      (try
+        (let [^ArrowFooter footer (.get arrow-footer-cache k
+                                        (fn [_] (Relation/readFooter (path->seekable-byte-channel path))))
+              blocks (.getRecordBatches footer)
+              ^ArrowBlock block (nth blocks block-idx nil)]
+          (if-not block
+            (throw (IndexOutOfBoundsException. "Record batch index out of bounds of arrow file"))
+            (util/with-open [arrow-buf (.get memory-cache (PathSlice. k (.getOffset block)
+                                                                      (+ (.getMetadataLength block) (.getBodyLength block)))
+                                             (fn [^PathSlice path-slice]
+                                               (let [buffer-cache-path (.resolve disk-store (.getPath path-slice))]
+                                                 (when-not (util/path-exists buffer-cache-path)
+                                                   (throw (os/obj-missing-exception k)))
+
+                                                 (CompletableFuture/completedFuture
+                                                  (Pair. (->resolved-path-slice path-slice buffer-cache-path) nil)))))]
+              (let [prefix-size (if (= (.getInt arrow-buf 0) MessageSerializer/IPC_CONTINUATION_TOKEN) 8 4)
```

Comment:
can this come out into a shared fn with the remote store? it's a pretty gnarly one to duplicate.

I think (although may be wrong) this is already written in Kotlin - if it's not quite there, would it be worth extracting there so that you can use it here?